Matrix host-configs and CI

bmhan12 opened the pull request:
This PR:

- Adds host-configs for matrix
  - Requires updating caliper version to a commit be compatible with newer CUDA.
- Updates uberenv commit
- (Tentatively/Temporarily) Adds workarounds for tests failing on matrix. Either switching from device to unified memory, or disablement when running with CUDA.

TODO:
- [x] Decide on cadence for matrix job (nightly? weekends? etc.)
- [x] Resolve or work around unit test failures

For now, jobs run every PR commit, but are allowed to fail.

bmhan12 commented:
For now, I have the matrix job set to run every PR commit, but allowed to fail. 
I will open up an issue tracking the current unit test failures and workarounds.

I've noticed even with the `wbronze` bank set as default for our runner, getting an allocation can still be difficult because the CI partition is overlapped with pbatch: https://lc.llnl.gov/confluence/spaces/GITLAB/pages/785285769/Special+partitions+for+CI+batch+jobs
(e.g. users can request pbatch nodes for the whole day, so no CI jobs will run)

As a byproduct of allowing the matrix job to fail, the matrix job timeout won't prevent PR's from merging.

kennyweiss reviewed (approved):
Thanks @bmhan12 

If you'd prefer to merge sooner, please create an issue to track the follow-up fixes we need to do for our cuda build, including the patches you added for umpire allocators and the failing tests.

bmhan12 commented on scripts/spack/configs/versions.yaml:
```diff
@@ -6,9 +6,10 @@ packages:
   caliper:
     require:
-      - spec: "@2.12.1"
+      - spec: "@git.527e1ae3a16487e425707e928e05c43d372e144f=develop"
```
I think just the CUDA host-configs are sufficient for now, I didn't notice any difference building & testing the non-cuda host-configs locally with newer caliper.

bmhan12 commented:
> If you'd prefer to merge sooner, please create an issue to track the follow-up fixes we need to do for our cuda build, including the patches you added for umpire allocators and the failing tests.

I added an issue #1743 to track these unit test failures, and added my thoughts on the errors and current workarounds if they were added.

publixsubfan commented:
Sorry, that looks like a typo on my end. I think the check should be `space == MemorySpace::Device`.

bmhan12 commented on src/axom/bump/extraction/TableBasedExtractor.hpp:
```diff
@@ -1170,6 +1170,8 @@ class TableBasedExtractor
       selectedZonesView.size(),
       AXOM_LAMBDA(axom::IndexType szIndex) {
+        // Avoid first-capture in constexpr-if context error
+        (void)selection;
```
Capture these variables outside of "if constexpr".

Documentation about this:
https://docs.nvidia.com/cuda/cuda-programming-guide/05-appendices/cpp-language-support.html#extended-lambda-restrictions
 
The exact rule line in question: "A variable cannot be implicitly captured inside an if-constexpr block that is lexically nested inside an extended lambda unless the variable has been implicitly captured outside the if-constexpr block or appears in the extended lambda’s explicit capture list."
